From b0877e884aa14b7a02c86fbaf21e8dd651a42d18 Mon Sep 17 00:00:00 2001
From: Zhen Lei <thunder.leizhen@huawei.com>
Date: Thu, 1 Sep 2022 16:57:21 +0800
Subject: [PATCH 059/261] [Huawei] arm64: kdump: Properly handle the 4G
 boundary

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187560

--------------------------------

The value of 'end' for both for_each_mem_range() and __map_memblock() is
'start + size', not 'start + size - 1'. So if the end value of a memory
block is 4G, then:
	if (eflags && (end >= SZ_4G)) {		//end=SZ_4G
		if (start < SZ_4G) {
			... ...
			start = SZ_4G;
		}
	}

	//Now, start=end=SZ_4G, all [start,...) will be mapped
	__map_memblock(pgdp, start, end, ..xxx..);

Fixes: e1ef167c3913 ("arm64: kdump: Don't force page-level mappings for memory above 4G")
Signed-off-by: Zhen Lei <thunder.leizhen@huawei.com>
---
 arch/arm64/mm/mmu.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/mm/mmu.c b/arch/arm64/mm/mmu.c
index e76765354040..a31f2124705e 100644
--- a/arch/arm64/mm/mmu.c
+++ b/arch/arm64/mm/mmu.c
@@ -523,13 +523,13 @@ static void __init map_mem(pgd_t *pgdp)
 			break;
 
 #ifdef CONFIG_KEXEC_CORE
-		if (eflags && (end >= SZ_4G)) {
+		if (eflags && (end > SZ_4G)) {
 			/*
 			 * The memory block cross the 4G boundary.
 			 * Forcibly use page-level mappings for memory under 4G.
 			 */
 			if (start < SZ_4G) {
-				__map_memblock(pgdp, start, SZ_4G - 1,
+				__map_memblock(pgdp, start, SZ_4G,
 					       pgprot_tagged(PAGE_KERNEL), flags | eflags);
 				start  = SZ_4G;
 			}
-- 
2.17.1

